File: frontend/src/lib/api/tryouts.ts
import { type IdParam, requestV2, toPathSegment } from "../api-client";
import type { RuntimeRunCreated } from "../types/runtime";
import type { TryoutExecuteInput, TryoutPersistRead, TryoutRead } from "../types/tryout";

function tryoutPath(runId: IdParam): string {
  return `/tryouts/${toPathSegment(runId)}`;
}

export function createTryout(
  payload: TryoutExecuteInput,
  signal?: AbortSignal,
): Promise<RuntimeRunCreated> {
  return requestV2<RuntimeRunCreated>("/tryouts", {
    body: payload,
    method: "POST",
    signal,
  });
}

export function getTryout(runId: IdParam, signal?: AbortSignal): Promise<TryoutRead> {
  return requestV2<TryoutRead>(tryoutPath(runId), { signal });
}

export function persistTryout(
  runId: IdParam,
  signal?: AbortSignal,
): Promise<TryoutPersistRead> {
  return requestV2<TryoutPersistRead>(`${tryoutPath(runId)}/persist`, {
    method: "POST",
    signal,
  });
}

export const tryoutsApi = {
  create: createTryout,
  get: getTryout,
  persist: persistTryout,
} as const;

File: backend/app/api/tryouts.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, status

from app.api.dependencies import get_tryout_service
from app.schemas.runtime import RuntimeRunCreated
from app.schemas.tryout import TryoutExecute, TryoutPersistRead, TryoutRead
from app.services.tryout_service import TryoutService

router = APIRouter(prefix="/tryouts", tags=["tryouts"])


@router.post("", response_model=RuntimeRunCreated, status_code=status.HTTP_201_CREATED)
def create_tryout(
    payload: TryoutExecute,
    service: Annotated[TryoutService, Depends(get_tryout_service)],
) -> RuntimeRunCreated:
    return service.create_tryout(payload)


@router.get("/{run_id}", response_model=TryoutRead)
def get_tryout(
    run_id: int,
    service: Annotated[TryoutService, Depends(get_tryout_service)],
) -> TryoutRead:
    return service.get_tryout(run_id)


@router.post("/{run_id}/persist", response_model=TryoutPersistRead)
def persist_tryout(
    run_id: int,
    service: Annotated[TryoutService, Depends(get_tryout_service)],
) -> TryoutPersistRead:
    return service.persist_tryout(run_id)
